pkg/gcrane/copy.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcrane
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/spf13/cobra"
+	"golang.org/x/sync/errgroup"
+)
+
+func init() { Root.AddCommand(NewCmdCopy()) }
+
+func NewCmdCopy() *cobra.Command {
+	recursive := false
+	cmd := &cobra.Command{
+		Use:     "copy",
+		Aliases: []string{"cp"},
+		Short:   "Efficiently copy a remote image from src to dst",
+		Args:    cobra.ExactArgs(2),
+		Run: func(cc *cobra.Command, args []string) {
+			doCopy(args, recursive)
+		},
+	}
+
+	cmd.Flags().BoolVarP(&recursive, "recursive", "r", false, "Whether to recurse through repos")
+
+	return cmd
+}
+
+func doCopy(args []string, recursive bool) {
+	src, dst := args[0], args[1]
+
+	c, err := newCopier(src, dst)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if recursive {
+		if err := c.recursiveCopy(src, dst); err != nil {
+			log.Fatalf("failed to copy images: %v", err)
+		}
+	} else {
+		if err := c.singleCopy(src, dst); err != nil {
+			log.Fatalf("failed to copy image: %v", err)
+		}
+	}
+}
+
+// Workaround for:
+// https://github.com/GoogleCloudPlatform/docker-credential-gcr/issues/54
+type staticAuth struct {
+	auth string
+	err  error
+}
+
+func (sa staticAuth) Authorization() (string, error) {
+	return sa.auth, sa.err
+}
+
+type copier struct {
+	srcRepo name.Repository
+	dstRepo name.Repository
+
+	srcAuth authn.Authenticator
+	dstAuth authn.Authenticator
+}
+
+func newCopier(src, dst string) (*copier, error) {
+	srcRepo, err := name.NewRepository(src, name.WeakValidation)
+	if err != nil {
+		if srcRef, refErr := name.ParseReference(src, name.WeakValidation); refErr == nil {
+			srcRepo = srcRef.Context()
+		} else {
+			return nil, fmt.Errorf("parsing repo %q: %v", src, err)
+		}
+	}
+
+	dstRepo, err := name.NewRepository(dst, name.WeakValidation)
+	if err != nil {
+		if dstRef, refErr := name.ParseReference(dst, name.WeakValidation); refErr == nil {
+			dstRepo = dstRef.Context()
+		} else {
+			return nil, fmt.Errorf("parsing repo %q: %v", dst, err)
+		}
+	}
+
+	srcHelper, err := authn.DefaultKeychain.Resolve(srcRepo.Registry)
```

Comment:
Why does gcrane use the default keychain?  Surely we can do better.

Reply:
Okay we actually can do better, this makes things a lot faster for listing without having to do the stupid static auth thing.

https://github.com/google/go-containerregistry/pull/315